src/yggdrasil/router.go

```diff
@@ -121,19 +125,39 @@ func (r *router) sendPacket(bs []byte) {
 		panic("Tried to send a packet shorter than a header...")
 	}
 	var sourceAddr address
-	var sourceSubnet subnet
-	copy(sourceAddr[:], bs[8:])
-	copy(sourceSubnet[:], bs[8:])
-	if !sourceAddr.isValid() && !sourceSubnet.isValid() {
-		return
-	}
 	var dest address
-	copy(dest[:], bs[24:])
 	var snet subnet
-	copy(snet[:], bs[24:])
-	if !dest.isValid() && !snet.isValid() {
+	var addrlen int
+	if bs[0]&0xf0 == 0x60 {
+		// IPv6 address
+		addrlen = 16
+		copy(sourceAddr[:addrlen], bs[8:])
+		copy(dest[:addrlen], bs[24:])
+		copy(snet[:addrlen/2], bs[24:])
+	} else if bs[0]&0xf0 == 0x40 {
+		// IPv4 address
+		addrlen = 4
+		copy(sourceAddr[:addrlen], bs[12:])
+		copy(dest[:addrlen], bs[16:])
+	} else {
+		// Unknown address length
+		return
+	}
+	if !r.cryptokey.isValidSource(sourceAddr, addrlen) {
 		return
 	}
+	if !dest.isValid() && !snet.isValid() {
+		if key, err := r.cryptokey.getPublicKeyForAddress(dest, addrlen); err == nil {
+			addr := *address_addrForNodeID(getNodeID(&key))
+			copy(dest[:], addr[:])
+			copy(snet[:], addr[:])
+			if !dest.isValid() && !snet.isValid() {
+				return
+			}
+		} else {
+			return
+		}
+	}
 	doSearch := func(packet []byte) {
 		var nodeID, mask *NodeID
```

Comment:
When creating a session for CKR routes, we should use the full key, not just the part we can see from the IP address. That would mean setting the nodeID and mask outside of `doSearch`. Similarly, for CKR routes, when we look up a session, we should check that the session matches the CKR key, not just the IP address.

Reply:
I've addressed this in fbfae47.